File: articles/genes-environment-and-developmental-plasticity/go/plasticity_summary.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

func col(header []string, name string) int {
	for i, h := range header {
		if h == name {
			return i
		}
	}
	return -1
}

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Usage: go run plasticity_summary.go data/genes_environment_plasticity_panel.csv")
	}
	f, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	rows, err := csv.NewReader(f).ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	if len(rows) < 2 {
		log.Fatal("CSV has no data rows")
	}

	header := rows[0]
	developmentIdx := col(header, "development_score")
	stressIdx := col(header, "embedded_stress")
	supportIdx := col(header, "embedded_support")
	careIdx := col(header, "current_care")
	currentStressIdx := col(header, "current_stress")
	if developmentIdx < 0 || stressIdx < 0 || supportIdx < 0 || careIdx < 0 || currentStressIdx < 0 {
		log.Fatal("Required columns missing")
	}

	count := 0
	developmentSum, stressSum, supportSum, careSum, currentStressSum := 0.0, 0.0, 0.0, 0.0, 0.0

	for _, row := range rows[1:] {
		development, e1 := strconv.ParseFloat(row[developmentIdx], 64)
		stress, e2 := strconv.ParseFloat(row[stressIdx], 64)
		support, e3 := strconv.ParseFloat(row[supportIdx], 64)
		care, e4 := strconv.ParseFloat(row[careIdx], 64)
		currentStress, e5 := strconv.ParseFloat(row[currentStressIdx], 64)
		if e1 != nil || e2 != nil || e3 != nil || e4 != nil || e5 != nil {
			continue
		}
		count++
		developmentSum += development
		stressSum += stress
		supportSum += support
		careSum += care
		currentStressSum += currentStress
	}

	fmt.Printf("Rows analyzed: %d\n", count)
	fmt.Printf("Mean development_score: %.4f\n", developmentSum/float64(count))
	fmt.Printf("Mean embedded_stress: %.4f\n", stressSum/float64(count))
	fmt.Printf("Mean embedded_support: %.4f\n", supportSum/float64(count))
	fmt.Printf("Mean current_care: %.4f\n", careSum/float64(count))
	fmt.Printf("Mean current_stress: %.4f\n", currentStressSum/float64(count))
}

File: articles/genes-environment-and-developmental-plasticity/cpp/plasticity_summary.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> values;
    std::stringstream ss(line);
    std::string item;
    while (std::getline(ss, item, ',')) values.push_back(item);
    return values;
}

int find_column(const std::vector<std::string>& header, const std::string& name) {
    for (size_t i = 0; i < header.size(); ++i) {
        if (header[i] == name) return static_cast<int>(i);
    }
    return -1;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: plasticity_summary data/genes_environment_plasticity_panel.csv\n";
        return 1;
    }

    std::ifstream file(argv[1]);
    if (!file.is_open()) {
        std::cerr << "Unable to open CSV file\n";
        return 1;
    }

    std::string line;
    std::getline(file, line);
    auto header = split_csv_line(line);

    int development_col = find_column(header, "development_score");
    int stress_col = find_column(header, "embedded_stress");
    int support_col = find_column(header, "embedded_support");
    int care_col = find_column(header, "current_care");
    int current_stress_col = find_column(header, "current_stress");

    long count = 0;
    double development_sum = 0.0, stress_sum = 0.0, support_sum = 0.0, care_sum = 0.0, current_stress_sum = 0.0;

    while (std::getline(file, line)) {
        auto values = split_csv_line(line);
        if ((int)values.size() <= development_col || (int)values.size() <= stress_col || (int)values.size() <= support_col || (int)values.size() <= care_col || (int)values.size() <= current_stress_col) continue;
        development_sum += std::stod(values[development_col]);
        stress_sum += std::stod(values[stress_col]);
        support_sum += std::stod(values[support_col]);
        care_sum += std::stod(values[care_col]);
        current_stress_sum += std::stod(values[current_stress_col]);
        count++;
    }

    std::cout << "Rows analyzed: " << count << "\n";
    std::cout << "Mean development_score: " << development_sum / count << "\n";
    std::cout << "Mean embedded_stress: " << stress_sum / count << "\n";
    std::cout << "Mean embedded_support: " << support_sum / count << "\n";
    std::cout << "Mean current_care: " << care_sum / count << "\n";
    std::cout << "Mean current_stress: " << current_stress_sum / count << "\n";
    return 0;
}
